package.json

```diff
@@ -18,19 +18,20 @@
   "types": "lib/peg.d.ts",
   "bin": "bin/peggy",
   "scripts": {
-    "pretest": "npm run parser",
-    "test": "npm run test:bare",
-    "test:bare": "mocha --recursive test --exclude test/browser.stub.js --reporter min",
+    "clean": "rimraf -f build && rimraf -f browser && mkdir build && cd build && mkdir ts && mkdir rollup && mkdir terser && cd .. && mkdir browser",
+    "terser": "terser -c passes=2 -m --sequences 40 build/rollup/peg.umd.js > build/terser/peggy.umd.min.js && terser -c passes=2 -m --sequences 40 --module build/rollup/peg.esmodule.js > build/terser/peggy.esmodule.min.js",
+    "rollup": "rollup -c",
+    "ts": "tsc --build tsconfig.json",
+    "set_version": "node -e \"const fs = require('fs'), pkg = JSON.parse(fs.readFileSync('./package.json')); fs.writeFileSync('./lib/version.js', '\\n\\'use strict\\';\\n\\nmodule.exports = \\'' + pkg.version + '\\';\\n');\"",
+    "test": "jest --coverage",
     "lint": "eslint . --ext .js bin/peggy",
-    "benchmark": "./benchmark/run",
-    "build": "npm run browser:build && npm run benchmark:build && npm run test:build",
-    "browser:clean": "rm -rf browser/",
-    "prebrowser:build": "npm run parser",
-    "browser:build": "browserify lib/peg.js -s peggy -o browser/peggy.min.js",
-    "postbrowser:build": "node tools/header browser/peggy.min.js && cp browser/peggy.min.js docs/vendor/peggy/peggy.min.js",
-    "benchmark:build": "browserify benchmark/browser.stub.js -o docs/js/benchmark-bundle.min.js && node tools/header docs/js/benchmark-bundle.min.js",
-    "test:build": "browserify test/browser.stub.js -o docs/js/test-bundle.min.js && node tools/header docs/js/test-bundle.min.js",
-    "parser": "node bin/peggy -o lib/parser.js --format commonjs src/parser.pegjs"
+    "benchmark": "node ./benchmark/run_bench.js",
+    "parser": "node bin/peggy -o lib/parser.js --format commonjs src/parser.pegjs",
+    "distro": "rm -f browser/peggy.min.js && cp build/terser/peggy.umd.min.js browser/peggy.min.js",
+    "audit": "npm run test && npm run benchmark",
+    "make": "npm run ts && npm run rollup && npm run terser && npm run parser",
+    "setup": "npm run clean && npm run set_version",
+    "build": "node ./bin/announce \"Setting up the build\" && npm run setup && node ./bin/announce \"Making libraries\" && npm run make && node ./bin/announce \"Running Audits\" && npm run audit && node ./bin/announce \"Creating distributions\" && npm run distro && node ./bin/announce \"Complete!\""
```

Comment:
The only one that probably needs to be copied is `peggy.min.js`, since one version goes into the distributed package and there needs to be one somewhere under `docs` for the website to use.

Reply:
> If this script is intended only for run on CI I prefer to implement this as steps in the CI config. That implementation would be much cleaner and maintainable. 

hard disagree.  this ties the library to the CI runner directly and means it cannot be built locally. 

&nbsp;

> Also, I do not see running linting during build. 

Correct, this is an omission on my part.  Fixing.

&nbsp;

> I suggest the following scenario in the GitHub Actions workflow, each as separate step with an appropriate name:
> 
> * npm run ts
> * npm run rollup
> * npm run terser
> * npm run lint
> * npm run test
> * npm run benchmark

Respectfully, absolutely not.  This would create a pipeline nightmare, slow the builds down to an undefined order cross-build invoking almost 120 individual builds, and make it literally not possible to build locally.

&nbsp;

> I do not see points in the creating files in one place and then copying to the another

This is standard behavior in nearly all build chains across nearly all languages for a reason.  You need to be able to inspect the intermediates.

&nbsp;

> so tasks distro and clean can be removed.

No, they cannot be.  Without these a build has not been cleaned properly, and may carry errors from the previous build.

Reply:
> there needs to be one somewhere under docs for the website to use.

Where would that go, please?  I didn't see that in the old build.

Reply:
`docs/vendor/peggy/peggy.min.js` which is not a great place, but where I found it when i took over.  If the location bothers you enough, please file another bug and we'll move it in a separate PR.

Reply:
I'll put it in the current location for now.  I don't much care either way, and if it does get moved, this PR isn't the right place to do it

Reply:
> Ideally, each step should be in buildable state, i.e. lint + build without errors

we do have that now, but it's being removed to fold the build steps into a single directory

Reply:
> Correct, this is an omission on my part. Fixing.

postponing the fix until after this lands because of the config file conflict underneath of #127

Reply:
> hard disagree. this ties the library to the CI runner directly and means it cannot be built locally.

Strange statement. Build is just `npm run rollup`. All other is packaging. Again, you can of course leave a common task calling all these scripts, do not call it in CI.

> Respectfully, absolutely not. This would create a pipeline nightmare, slow the builds down to an undefined order cross-build invoking almost 120 individual builds, and make it literally not possible to build locally.

Hmm? Where did the number 120 come from? And where is pipeline nightmare? That is the absolutely the same thing as calling commands via `... && ...` but with clear grouping/folding from the CI (even with time measurements of each step!). You yourself saw how to simply miss some step in a long unreadable line, while in an explicit list making such a mistake is much more difficult.

> This is standard behavior in nearly all build chains across nearly all languages for a reason. You need to be able to inspect the intermediates.

I'm not saying that you shouldn't keep intermediate results. Just why copy them? Why not create immediately in the right place?

> No, they cannot be. Without these a build has not been cleaned properly, and may carry errors from the previous build.

CI will always operate on the fresh checkout, so there is nothing to clean. You can keep it for manual builds although, but I think that all used instruments will overwrite existing files by default without any notice so it seems pretty useless for me... but all people different, so I don't mind if you're so comfortable

Reply:
> Strange statement. Build is just npm run rollup. All other is packaging. Again, you can of course leave a common task calling all these scripts, do not call it in CI.

Go ahead and do the work you're requesting yourself, on a fork.  Once you've run it you'll see the problem.

I'm not interested in doing the labor to support your idea.

&nbsp;

> Hmm? Where did the number 120 come from?

The number of steps you're requesting multiplied by the number of runner jobs.

&nbsp;

>  And where is pipeline nightmare? 

Go ahead and do the work you're requesting yourself, on a fork.  Once you've run it you'll see the problem.

What you're describing suggests that you've never actually set up a chained build under this system, and don't know how it works.

.

> That is the absolutely the same thing as calling commands via ... && ... but with clear grouping/folding from the CI

No, it isn't.

You don't even get artifacts between steps, dude.

Go ahead and do the work you're requesting yourself, on a fork.  Once you've run it you'll see the problem.

&nbsp;

&nbsp;

> I'm not saying that you shouldn't keep intermediate results. Just why copy them? 

I already explained this to you several times.

1. Because you want all of the intermediates on disk when you're debugging them.
2. Because this allows partial analysis
3. Because it's slightly more convenient, since the tools use the same names each time, so I don't have to manually rename them, or de-collide them
4. Because it's more work to clean them up inbetween
    1.  As evidence, I already fixed this for you, because you keep complaining about it and won't do the work yourself.
    2. If you would try to do the work yourself, you'd notice that this is now `npm run tiny`
    3. It's extra work, not saved work
    4. This is completely meaningless.  You're just creating work for me.  There is no technical upside
5. Because cleaning them up after is silly, but I did it anyway, because it's getting hard to find the technical comments inbetween all this [bikeshedding](https://en.wiktionary.org/wiki/bikeshedding)

Reply:
Please use this PR for comments with a specific technical improvement, moving forwards.  Thank you

Reply:
Please actually fork the library, @Mingun , and create the github action you keep asking me to create.  

After that, get it running locally.

I would like you to understand:

1. The change you're requesting either means it doesn't work locally or it works fundamentally differently remotely
2. If CI works differently, it's no longer a valid test
3. You now have to install two other github actions - upload-artifact and download-artifact - then configure them
4. The builds are actually quite different build platform to build platform, because of stubs and versions
    1. If you build this on `node 15` you'll get quite a different output than on `node 10`
    2. This is true of `babel` too
5. The configuration in this case is brutal
    1. You have to have a special case for each build, because of naming
    2. The various subordinate builds will have different shims and shams inside
    3. They as such aren't exchangeable
    4. They won't persist step to step.  You have to put them in side storage
    5. That has to be done ***in the yaml*** because you need to know which platform you're on so it can have a different name
    6. This has to be done twice at every step except the first and the last, to download the previous and upload the next
    7. You're doing this over six steps (two short) over 20 builds, so that's (12-1-1) 200 uploads and downloads
    8. That's not fast, and you have a monthly cap, so now the builds will just magically break some months for no apparent reason, and it takes a GH Actions person to figure out why
    9. This also means that every time we add a node version, we have to remember to go change this, and fix the build matrix
6. Afterwards I want you to compare the current speed (about three minutes) to the new speed
    1. Best guess you'll hit eight minutes
7. Next look at how much work you had to do
    1. Did you enjoy doing it?
    2. It's around 150 lines of YAML, because the two actions have to be configured per-platform
    3. Gee, YAML's easy, isn't it?  I bet you really enjoyed learning that the quoting rules are different in differing contexts
    4. At least you only had to create four new YAML files in addition to tripling the length of the one we already have
    5. Besides it's just two new completely unnecessary maintenance tasks, and a complete separation between how CI works and how a local build works
    6. Being night and day different from the basic work is the definition of a good test system, isn't it?  

The next time you have an idea, please don't block my work while asking me to implement it for you.  Thank you.

Reply:
"why don't you just leave it on disk"  because you have about a 5% chance of losing the image, and at 200 steps, that means you lose four in an average build.

that means that, on average, you need to run 32 times to get past the flakiness, unless you waste a ton of effort uploading and downloading fragile, rate limited artifacts, which will bury the artifacts we really want to keep.